### DISCUSSION TITLE
Most straightforward way to add RDKit postgres cartridge to an existing posgresql?

### DISCUSSION THREAD

**Discussion by drassokhin** [rdkit/rdkit#5940] [Q&A]
I have a running and fully operational PostgreSQL (PostgreSQL 15.1 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-15), 64-bit) plus libraries and headers for C language development, so I have been able to build and install some other postgres extensions. Would you be recommending building and installing the RDKit cartridge from source? Is the latest version compatible with PG 15? Will ```make``` / ```make install``` with the Makefile found in https://github.com/rdkit/rdkit/tree/master/Code/PgSQL/rdkit and then running ```CREATE EXTENSION ...``` to load it into a database work?

BTW, I also tried the conda install (just to see how the conda build works)
```
conda create -c conda-forge -n pg-rdkit-env rdkit
conda activate pg-rdkit-env
conda install -c rdkit rdkit-postgresql
```
and it failed with UnsatisfiableError.

**Comment by ptosco** [comment]
@drassokhin I have not tried myself building against PostgreSQL 15, but it most likely is.
That `Makefile` is obsolete; the PostgreSQL cartridge build is integrated with the CMake build system; you will only need to add the relevant CMake command line switches to enable the cartridge build. Please find below my own CMake command line switches as reference; you should only need to change the paths to match your Postgres 15 installation:
```
-D RDK_BUILD_PGSQL=ON \
-D RDK_PGSQL_STATIC=ON \
-D PostgreSQL_INCLUDE_DIR=/scratch/toscopa1/install/postgres_13_1/include \
-D PostgreSQL_TYPE_INCLUDE_DIR=/scratch/toscopa1/install/postgres_13_1/include/server \
-D PostgreSQL_LIBRARY=/scratch/toscopa1/install/postgres_13_1/lib/libpq.so \
```
When you run `make install`, you will then see a message regarding the PostgreSQL build; if it fails, the message will tell you how to complete the installation after gaining `root` privileges to install the cartridge in the `/usr` directory tree.
You will then need to restart the `postgres` service before running `CREATE EXTENSION`.

**Reply by drassokhin** [reply]
Hi Paulo @ptosco!
Thanks for your prompt response! I will give the CMake a try, as you recommended. 
I just want to build the cartridge w/o RDKit Python wrappers and run some quick tests. 

BTW, I have recently been playing with "similarity searching", not necessarily in chemical structure spaces, but in various embedding vector spaces where a "distance" or "similarity" can be introduced. I was really impressed by the performance of the built-in bit operations in the recent versions of PostgreSQL. For example, for a test, I have created a table containing a column of the ```bit(1024)``` type, filled it with fingerprints of the entire ChEMBL 31 (2.3M structures; fingerprints computed with RDKit ```GetMorganFingerprintAsBitVect```) and found out that even the simplest brute force similarity search with a full table scan and Tanimoto similarity defined as ```((bit_count(((b1)::"bit" & (b2)::"bit")))::double precision / (bit_count(((b1)::"bit" | (b2)::"bit")))::double precision)``` takes less than a second on a 'typical' multi-processor AWS EC2 instance.

**Reply by xuchunqiu** [reply]
I found `make install` is unnecessary if just want to use RDKit cartridge. Did I get it wrong?

First, `cmake`.

```bash
# if I didn't build python part, cmake will get error
$ cmake \
-DCMAKE_CXX_COMPILER=/opt/rh/devtoolset-8/root/usr/bin/g++ \
-DRDK_BUILD_PGSQL=ON \
-DPostgreSQL_ROOT="/usr/pgsql-15" \
-DPostgreSQL_LIBRARY="/usr/pgsql-15/lib/libpq.so" \
-DPostgreSQL_INCLUDE_DIR=/usr/pgsql-15/include/server \
-DPostgreSQL_TYPE_INCLUDE_DIR=/usr/pgsql-15/include/server/catalog \
-DRDK_BUILD_PYTHON_WRAPPERS=ON \
-DPYTHON_EXECUTABLE="/usr/local/python3/bin/python3.9" \
-DPYTHON_NUMPY_INCLUDE_PATH=/usr/local/python3/lib/python3.9/site-packages/numpy/core/include \
-DPYTHON_INCLUDE_DIR=/usr/local/python3/include/python3.9

-- Found Catch2 source in /home/username/rdkit-build/rdkit-Release_2023_09_1/External/catch/catch
CATCH: /home/username/rdkit-build/rdkit-Release_2023_09_1/External/catch/catch/single_include/catch2
-- Boost 1.58.0 found.
-- Found Boost components:
   python3
PYTHON Py_ENABLE_SHARED: 0
PYTHON USING LINK LINE: -pthread -shared
nbval not found, disabling the jupyter tests
-- Boost 1.58.0 found.
-- Found Boost components:
   system;serialization;iostreams
== Using strict rotor definition
-- Boost 1.58.0 found.
-- Found Boost components:
   system;iostreams
-- maeparser include dir set as '/usr/local/include'
-- maeparser libraries set as '/usr/local/lib64/libmaeparser.so'
-- coordgen include dir set as /usr/include
-- coordgen libraries set as '/usr/lib64/libcoordgen.so'
-- Found RingDecomposerLib source in /home/username/rdkit-build/rdkit-Release_2023_09_1/External/RingFamilies/RingDecomposerLib/src/RingDecomposerLib
-- Boost 1.58.0 found.
-- Found Boost components:
   iostreams
-- Boost 1.58.0 found.
-- Found Boost components:
   system;iostreams
/home/username/rdkit-build/rdkit-Release_2023_09_1/Data/Fonts/ComicNeue-Regular.ttf already there, skipping the download
-- Boost 1.58.0 found.
-- Found Boost components:
   program_options
-- Found RapidJSON source in /home/username/rdkit-build/rdkit-Release_2023_09_1/External
postgres: /usr/pgsql-15/include/server;/usr/pgsql-15/include/server/catalog
-- Configuring done
-- Generating done
-- Build files have been written to: /home/username/rdkit-build/rdkit-Release_2023_09_1

# got
Code/PgSQL/rdkit/pgsql_install.sh
Code/PgSQL/rdkit/pgsql_regress.sh
Code/PgSQL/rdkit/update_sql/*.sql
```

`Code/PgSQL/rdkit/pgsql_install.sh` is the file that install the PostgreSQL RDKit cartridge according to [Code/PgSQL/rdkit/README.md](https://github.com/rdkit/rdkit/blob/master/Code/PgSQL/rdkit/README.md). It
just copy files to `PostgreSQL_ROOT`. Currently it only missing `Code/PgSQL/rdkit/librdkit.so`.

```sh
set -x
test -n "$DESTDIR" && \
  mkdir -p "$DESTDIR/usr/pgsql-15/share/extension" && \
  mkdir -p "$DESTDIR/usr/pgsql-15/lib/"
cp "/home/username/rdkit-build/rdkit-Release_2023_09_1/Code/PgSQL/rdkit/rdkit--4.4.0.sql" "$DESTDIR/usr/pgsql-15/share/extension"
cp "/home/username/rdkit-build/rdkit-Release_2023_09_1/Code/PgSQL/rdkit/rdkit.control" "$DESTDIR/usr/pgsql-15/share/extension"
cp "/home/username/rdkit-build/rdkit-Release_2023_09_1/Code/PgSQL/rdkit/update_sql/"*.sql "$DESTDIR/usr/pgsql-15/share/extension"
cp "/home/username/rdkit-build/rdkit-Release_2023_09_1/Code/PgSQL/rdkit/librdkit.so" "$DESTDIR/usr/pgsql-15/lib/rdkit.so"
```

After `make` we already have `librdkit.so`.

```bash
$ make -j$(nproc)
...
[ 99%] Built target testRGroupDecompInternals
[100%] Linking CXX executable gaExample
[100%] Built target gaExample
[100%] Linking CXX executable rgroupCatchTests
[100%] Built target rgroupCatchTests
[100%] Linking CXX shared module ../../../../rdkit/Chem/rdRGroupDecomposition.so
[100%] Built target rdRGroupDecomposition
[100%] Linking CXX shared module ../../../../rdkit/Chem/rdSubstructLibrary.so
[100%] Built target rdSubstructLibrary
[100%] Linking CXX executable testRGroupDecomp
[100%] Built target testRGroupDecomp

# got
Code/PgSQL/rdkit/librdkit.so

# run pgsql_install.sh
$ sudo systemctl stop postgresql-15
$ sudo sh Code/PgSQL/rdkit/pgsql_install.sh
$ sudo systemctl start postgresql-15

# test
$ sudo -i -u postgres
$ psql -p 12345
$ createdb -h localhost -p 12345 nci5K
$ psql -c 'create extension rdkit' -h localhost -p 12345 nci5K
CREATE EXTENSION
$ psql -c "select mol_to_smiles(mol_from_smiles('O1OCCCC1'));" -h localhost -p 12345 nci5K
 mol_to_smiles
---------------
 C1CCOOC1
(1 row)
```

**Moreover, `make install` will get a file not found error on `rdCoordGen.so`, which is same as https://github.com/rdkit/rdkit/issues/2468.**

```bash
$ make install
...
[ 99%] Built target rdChemicalFeatures
[100%] Built target rdkit
Install the project...
-- Install configuration: "Release"
-- Installing: /home/username/rdkit-build/rdkit-Release_2023_09_1/lib/cmake/rdkit/rdkit-targets.cmake
-- Installing: /home/username/rdkit-build/rdkit-Release_2023_09_1/lib/cmake/rdkit/rdkit-targets-release.cmake
-- Up-to-date: /home/username/rdkit-build/rdkit-Release_2023_09_1/lib/cmake/rdkit/rdkit-config.cmake
-- Up-to-date: /home/username/rdkit-build/rdkit-Release_2023_09_1/lib/cmake/rdkit/rdkit-config-version.cmake
CMake Error at External/CoordGen/Wrap/cmake_install.cmake:55 (file):
  file INSTALL cannot find
  "/home/username/rdkit-build/rdkit-Release_2023_09_1/rdkit/Chem/rdCoordGen.so".
Call Stack (most recent call first):
  External/CoordGen/cmake_install.cmake:42 (include)
  External/cmake_install.cmake:45 (include)
  cmake_install.cmake:93 (include)


make: *** [install] Error 1

# although I can workaround this by:
$ make  # generate rdCoordGen.so
$ sudo chattr +i /home/username/rdkit-build/rdkit-Release_2023_09_1/rdkit/Chem/rdCoordGen.so  # make it undeletable by make install
$ awk 'NR>=58 && NR<=61 { if ($0 !~ /^#/) print "#" $0; else print $0; next } 1' External/CoordGen/Wrap/cmake_install.cmake > tmp_file && mv tmp_file External/CoordGen/Wrap/cmake_install.cmake  # comment out line 58-61
$ make install
# but there will encounter same error w/ libRDKitRDBoost.so.1.2023.09.1, rdBase.so, libRDKitDataStructs.so.1.2023.09.1, ...
```

**Comment by dwilbur-sutrobio** [comment]
is there any chance that this could be added to the documentation as a step by step recipe for vs people guessing what they need to do?

**Comment by rccc** [comment]
Installation procedure on existing postgres is not clear / not complete.

conda install -c rdkit rdkit-postgresql

psql -c 'CREATE EXTENSION rdkit' db    ->    "mol' type already exists

psql -c 'DROP EXTENSION rdkit CASCADE db   ->   "extention does not exists"

Such a waste of time !

**Comment by rccc** [comment]
You will not be able to install this cartbridge on existing database. Everyone should consider this cartbridge as deprecated as there is no support from RDKIT community and documentation is not maintained.

I advise everyone to use Bingo from EPAM instead. The difference is that this cartbridge is maintained by paid developpers.

**Reply by ptosco** [reply]
@rccc Please **do not provide incorrect information**.
It is of course possible to install the PostgreSQL cartridge on an existing PostgreSQL installation.
The cartridge is neither deprecated nor unmaintained.
A few days ago I have updated the [documentation to install the cartridge](https://github.com/rdkit/rdkit/blob/master/Code/PgSQL/rdkit/README.md) as mentioned in [another discussion](https://github.com/rdkit/rdkit/discussions/6182).

**Comment by dwilbur-sutrobio** [comment]
then... could I request that the conda repo gets updated to postgres 14.  v12 is getting pretty old, and all our other postgres databases are at v14 (its the default postgres version for an ubuntu 22.04).  likewise I really would like to see the doc updated on how to do that on a pre existing postgres or non conda provided rdkit-postgres. I will review your additions since they are relevant to that. (thanks for working on that)

**Reply by ptosco** [reply]
@dwilbur-sutrobio This `README` explains how to build and install the cartridge on an existing PostgreSQL installation:
https://github.com/rdkit/rdkit/blob/master/Code/PgSQL/rdkit/README.md
I will also update https://www.rdkit.org/docs/Cartridge.html as soon as I find some time.

**Reply by dwilbur-sutrobio** [reply]
many thanks =)

**Comment by dwilbur-sutrobio** [comment]
also is GitHub replacing where to look for rdkit documentation verses here https://www.rdkit.org/docs/Cartridge.html ?
